### ISSUE TITLE
push adds a default, unique tag

### ISSUE THREAD

**Issue by seemiller** [carvel-dev/imgpkg#248]
**Describe the problem/challenge you have**

OCI registries can perform a garbage collection process where any un-tagged image is deleted.

This can cause a problem when tags are re-used, with the tag migrating from an older image to a newer one, but with the digest of the older image still being referenced.

Here's how the problem arises.

1. An image with URL/Digest `image@sha256:ABC123` is tagged `v.1.2.3`.
2. Image is referenced in the package repository with the digest `image@sha256:ABC123`
3. The new image uploaded, gets the digest, `image@sha256:ZCYD23`
4. The tag is move from `ABC123` to `ZCYD23`
5. `image@sha256:ABC123` _no longer has a tag_.
5. Package repository still points to `image@sha256:ABC123`
6. Garbage collector deletes `image@sha256:ABC123`
7. Users still using the original package repo will now experience `image not found` errors when attempting to install the package or package repository.

![Registry GC](https://user-images.githubusercontent.com/5900/133817744-6e8bcc27-acd3-4691-ad71-afafdc4a2e58.png)

**Describe the solution you'd like**

imgpkg should provide a means to add a unique/arbitrary tag when pushing images. Any image with this unique/arbitrary tag would never be garbage collected unless that tag was manually removed.

This would be optional and driven by a command line argument or environment variable.

`imgpkg push --bundle registry.example.com/foo/image:v1.2.3 --file ./bundle --unique-tag`

**Anything else you would like to add:**
[Additional information that will assist in solving the issue.]

---
Vote on this request

This is an invitation to the community to vote on issues, to help us prioritize our backlog. Use the "smiley face" up to the right of this comment to vote.

👍 "I would like to see this addressed as soon as possible"
👎 "There are other more important things to focus on right now"

We are also happy to receive and review Pull Requests if you want to help working on this issue.

**Comment by jorgemoralespou** [comment]
@seemiller I'm assuming the `unique-tag` will be in addition to the provided tag for the repository. 

- Any recommendation on the tag format?
- Should it be configurable, or default is ok?

**Comment by seemiller** [comment]
At the moment, I have no opinion on the format of the tag. Options include:
* A (partial) SHA
* The time stamp of the upload
* Randomly generated string
* A user provided value
* `fred`

I'm not to attached to any particular format as all that is needed is another tag to prevent garbage collection.

**Comment by cppforlife** [comment]
this shouldnt be a dedicated flag since all users should benefit from this (and for consistency with impgkg copy). tag should be standard imgpkg uploading tag eg `sha256-abc...123.imgpkg` (https://github.com/vmware-tanzu/carvel-imgpkg/blob/7e15cf608460ca3e145e21773d41366c4e44e022/pkg/imgpkg/imageset/image_set.go#L215). (as a side note that standard convention helps folks to apply harbor replication rules if they use that to copy things instead of imgpkg copy).

**Comment by DennisDenuto** [comment]
@seemiller as discussed in the [carvel community meeting](https://hackmd.io/F7g3RT2hR3OcIh-Iznk2hw) this sounds like a feature we would like to support in imgpkg

I agree with @cppforlife that we should use a standardized tag name (similar to what happens in copy)

I'll go ahead and mark this as carvel-accepted, meaning we plan on working on this.

**Comment by jvalkeal** [comment]
I would really like to define all my tags with `push` command. If I later pull bundle to add tag(s), digest in a repository will change. For example:
```
export TAG=2.9.0-SNAPSHOT
export RTAG=$(date +%y%m%d%s%N)

imgpkg push --bundle jvalkeal/scdf-package:$RTAG --file templates/generated/package/scdf/bundle
docker pull jvalkeal/scdf-package:$RTAG
docker tag jvalkeal/scdf-package:$RTAG jvalkeal/scdf-package:$TAG
docker tag jvalkeal/scdf-package:$RTAG jvalkeal/scdf-package:$TAG-$RTAG
docker push jvalkeal/scdf-package:$TAG
docker push jvalkeal/scdf-package:$TAG-$RTAG
```
![image](https://user-images.githubusercontent.com/50398/137162980-85b52caa-8b2c-4b4d-8333-6fdc0d087a0f.png)

**Comment by aaronshurley** [comment]
@seemiller are you still planning to work on this issue? Any challenges so far?

**Comment by seemiller** [comment]
I am, was hoping to start last week, but KubeCon kinda got in the way!

**Comment by DennisDenuto** [comment]
@jvalkeal interesting! Looks like docker is re-formatting the manifest json before pushing it to the registry, adding new lines in the json (hence the digest change)

As a workaround I would recommend using crane to tag the image (crane uploads the tag object without messing with the manifest json)

wrt specifying multiple tags, this issue seems to be scoped to fixing images being gc'd incorrectly. Do you mind creating a separate issue to capture the use case around specifying multiple tags for bundles / images?

**Comment by seemiller** [comment]
Discussed this issue with @DennisDenuto and @joaopapereira

- Instead of optionally adding a tag, imgpkg will _always_ add it
- Tag format will be consistent with what is currently done in the copy cmd - `{algo}-{hex}.imgpkg`
- Implement new code in the `Push` function in `contents.go`
- Add test coverage in `image_test.go` and `bundle_test.go`

**Comment by joaopapereira** [comment]
Released on version 0.24.0
